From 9278337ebf510c067f4aa468411b4e7a7b9804b3 Mon Sep 17 00:00:00 2001
From: "sheng.liu" <sheng.liu@amlogic.com>
Date: Wed, 10 Nov 2021 14:06:17 +0800
Subject: [PATCH] gstreamer1.0-plugins-good: add env value to set dw mode in
 v4l2  [1/1]

PD#SWPL-63603

Problem:
There is no way to set dw mode in v4l2

Solution:
Add V4L2_SET_AMLOGIC_DW_MODE env value to set dw mode

Verify:
AP222
---
 sys/v4l2/gstv4l2object.c | 39 ++++++++++++++++++++++++++++++++++++---
 1 file changed, 36 insertions(+), 3 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 8bb0755..4347963 100755
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -3255,6 +3255,41 @@ gst_v4l2_video_colorimetry_matches (const GstVideoColorimetry * cinfo,
   return FALSE;
 }
 
+static void
+set_amlogic_vdec_parm(GstV4l2Object *v4l2object, struct v4l2_streamparm *streamparm)
+{
+  struct aml_dec_params *decParm= (struct aml_dec_params*)streamparm->parm.raw_data;
+  const char *env;
+
+  decParm->cfg.metadata_config_flag = 1 << 13;
+
+  if (v4l2object->type == V4L2_BUF_TYPE_VIDEO_OUTPUT
+      || v4l2object->type == V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE)
+  {
+    env= getenv("VENDOR_MEDIA_GST_VIDEO_SET_DW_MODE");
+    if ( env )
+    {
+      int dwMode= atoi(env);
+      switch( dwMode )
+      {
+         case 0: case 1: case 2: case 3: case 4: case 16:
+            decParm->cfg.double_write_mode= dwMode;
+            decParm->parms_status |= V4L2_CONFIG_PARM_DECODE_CFGINFO;
+            break;
+      }
+    }
+
+    if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_S_PARM, streamparm) < 0)
+    {
+      GST_DEBUG_OBJECT (v4l2object->dbg_obj, "set vdec parm fail");
+    }
+    else
+    {
+      GST_DEBUG_OBJECT (v4l2object->dbg_obj, "Set dwMode to %d", decParm->cfg.double_write_mode);
+    }
+  }
+}
+
 static gboolean
 gst_v4l2_object_set_format_full (GstV4l2Object * v4l2object, GstCaps * caps,
     gboolean try_only, GstV4l2Error * error)
@@ -3285,6 +3320,10 @@ gst_v4l2_object_set_format_full (GstV4l2Object * v4l2object, GstCaps * caps,
   if (!try_only)
     GST_V4L2_CHECK_NOT_ACTIVE (v4l2object);
 
+  memset (&streamparm, 0x00, sizeof (struct v4l2_streamparm));
+  streamparm.type = v4l2object->type;
+  set_amlogic_vdec_parm(v4l2object, &streamparm);
+
   is_mplane = V4L2_TYPE_IS_MULTIPLANAR (v4l2object->type);
 
   gst_video_info_init (&info);
@@ -3660,9 +3699,6 @@ gst_v4l2_object_set_format_full (GstV4l2Object * v4l2object, GstCaps * caps,
   GST_DEBUG_OBJECT (v4l2object->dbg_obj, "Desired framerate: %u/%u", fps_n,
       fps_d);
 
-  memset (&streamparm, 0x00, sizeof (struct v4l2_streamparm));
-  streamparm.type = v4l2object->type;
-
   if (v4l2object->ioctl (fd, VIDIOC_G_PARM, &streamparm) < 0)
     goto get_parm_failed;
 
-- 
2.32.0

